cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowContext.java

```diff
@@ -35,16 +47,38 @@
   private final ProgramWorkflowRunner programWorkflowRunner;
   private final Map<String, String> runtimeArgs;
   private final WorkflowToken token;
+  private final Metrics userMetrics;
 
-  BasicWorkflowContext(WorkflowSpecification workflowSpec, @Nullable WorkflowActionSpecification specification,
+  BasicWorkflowContext(WorkflowSpecification workflowSpec, @Nullable WorkflowActionSpecification spec,
                        long logicalStartTime, @Nullable ProgramWorkflowRunner programWorkflowRunner,
-                       Map<String, String> runtimeArgs, WorkflowToken token) {
+                       Map<String, String> runtimeArgs, WorkflowToken token, Program program, RunId runId,
+                       MetricsCollectionService metricsCollectionService,
+                       DatasetFramework datasetFramework, DiscoveryServiceClient discoveryServiceClient) {
+    super(program, runId, new BasicArguments(runtimeArgs), (spec == null) ? new HashSet<String>() : spec.getDatasets(),
+          getMetricCollector(program, runId.getId(), metricsCollectionService),
+          datasetFramework, discoveryServiceClient, null, null);
     this.workflowSpec = workflowSpec;
-    this.specification = specification;
+    this.specification = spec;
     this.logicalStartTime = logicalStartTime;
     this.programWorkflowRunner = programWorkflowRunner;
     this.runtimeArgs = ImmutableMap.copyOf(runtimeArgs);
     this.token = token;
+    if (metricsCollectionService != null) {
+      this.userMetrics = new ProgramUserMetrics(getProgramMetrics());
+    } else {
+      this.userMetrics = null;
+    }
+  }
+
+  @Nullable
+  private static MetricsContext getMetricCollector(Program program, String runId,
```

Comment:
Better have this method returns `Metrics` so that you don't need to check for null for `MetricsCollectionService` in both constructor and this method (you just need to check here).